### ISSUE TITLE
Add unit tests to compare log splitter analysis option vs. current triggers

### ISSUE THREAD

**Issue by wexxlee** [OverlayPlugin/cactbot#94]
### Feature(s) Requested

Per comment from @valarnin on #92:

> Another good follow up PR would be some unit testing checking that this is getting all the lines we currently use in triggers, by checking currently used triggers or something, but that's a bit more complex. (basically import all the triggers, loop through them to get line types used, construct default versions of those line types via the log type definitions, then passing those through the splitter and checking the output)

Definitely supportive of this.  In terms of implementation, we probably need to give some thought to where that unit test should live/how it should be run.  If it's folded into the current `mocha` test bed, that means that anyone PR'ing triggers that use a new logline type for the first time would cause the test workflow job to fail (and if merged, would cause all future PRs/merges to similarly fail that workflow) until log splitter is updated.  Maybe that's okay if we want to enforce it to that degree?  Just starting the discussion though...

**Comment by JLGarber** [comment]
I sort of wonder whether it might make sense to have the list of implemented log lines live outside the splitter as a resource file. This way we could have both this test and the splitter import it and have to update it only once at any time. It would also be a lot simpler to point new people at a resource file that needs updating vs "oh, this random utility breaks stuff". Maybe I'm overthinking it, but we want to avoid these kinds of blockers as much as possible given the expansion release coming up.

**Comment by wexxlee** [comment]
> I sort of wonder whether it might make sense to have the list of implemented log lines live outside the splitter as a resource file. This way we could have both this test and the splitter import it and have to update it only once at any time. It would also be a lot simpler to point new people at a resource file that needs updating vs "oh, this random utility breaks stuff". Maybe I'm overthinking it, but we want to avoid these kinds of blockers as much as possible given the expansion release coming up.

Building on that, instead of having people ever manually edit the resource file, we can create it one-time and then have a workflow that runs only on merge commits to `main` that looks for new trigger types and, if one is found, adds it to the resource file (well, I suppose, opens a PR to add it to the resource file, so we don't have a workflow merging unreviewed commits directly).  That would solve the need for anyone to actively keep the resource file in sync.

If we wanted to do unit tests on top of that, we can create a unit test script and hook it to mocha.  It would automatically run against the auto-created PR to tell us if there's an issue with the changed resource file before it gets merged.

**Comment by JLGarber** [comment]
Yes, now that you write it out like that I'm not sure why I didn't just go to that immediately. Automate as much as possible.

I agree that we should have the automated workflow stop at generating the PR. That's still going to be a soft blocker for any contributors who might happen to use a new line type for the first time, but it's extremely unlikely that contributor will be new to the repository. If it happens to any of us who contribute regularly, it shouldn't be much of a burden to jump over and approve the auto-generated PR.

**Comment by wexxlee** [comment]
> That's still going to be a soft blocker for any contributors who might happen to use a new line type for the first time, but it's extremely unlikely that contributor will be new to the repository. If it happens to any of us who contribute regularly, it shouldn't be much of a burden to jump over and approve the auto-generated PR.

The way I'm envisioning it, I think we could even remove the blocking element.  So the flow would be something like:

1. `used_log_types` resource file contains all currently used logline types.  It gets imported/used by `splitter.ts` at the time someone splits a log with the analysis flag/checkbox turned on.  It's also read by a `mocha` test to determine which line types to unit test against the splitter function.
2. Any contributor opens a PR with a trigger that uses a new log type.  (The `mocha` test workflow continues to run fine, since nothing new has been added to `used_log_types` yet.
3. The PR with triggers is approved & merged.
4. The merging of the PR fires an `update_usedlogtypes` workflow that runs a script to check all log types currently in use in the repo, and updates `used_log_types` if any new ones are found. 
5. If that results in changes to `used_log_types`, the workflow will create a new branch, commit the changed `used_log_types` file, and open a PR against the new branch.  (Note: This step can be handled entirely by an existing GH action like [this one](https://github.com/peter-evans/create-pull-request).
6. The PR being opened will -- as normal with PRs -- trigger the test workflow that runs `mocha`.  It automatically runs against the PR head branch, so the updated `used_log_types` file will be used.  If the update results in a failed unit test, it'll get flagged on the PR; if not, the PR can be merged at any point (it won't block anything in the meantime).

Takes a lot of words to write out, but I think it's pretty clean, implementation-wise?

**Comment by wexxlee** [comment]
Closing this as done via #128.  There was a possible to-do relating to adding unit testing to ensure the filter was correctly processing all the lines it was supposed to, but given the way 109/128 were implemented and as I mentioned [here](https://github.com/OverlayPlugin/cactbot/issues/125#issuecomment-2045608688), I don't think there's much else that can be done for unit testing.
